routing/road_graph.hpp

```diff
@@ -99,17 +111,74 @@ class IRoadGraph
   };
 
   /// This class is responsible for loading edges in a cross.
-  /// It loades only outgoing edges.
-  class CrossEdgesLoader
+  class ICrossEdgesLoader
   {
   public:
-    CrossEdgesLoader(m2::PointD const & cross, TEdgeVector & outgoingEdges);
+    ICrossEdgesLoader(m2::PointD const & cross, IRoadGraph::Mode mode, TEdgeVector & edges)
+      : m_cross(cross), m_mode(mode), m_edges(edges)
+    {
+    }
+    virtual ~ICrossEdgesLoader() = default;
 
-    void operator()(FeatureID const & featureId, RoadInfo const & roadInfo);
+    void operator()(FeatureID const & featureId, RoadInfo const & roadInfo)
+    {
+      LoadEdges(featureId, roadInfo);
+    }
 
   private:
+    virtual void LoadEdges(FeatureID const & featureId, RoadInfo const & roadInfo) = 0;
+
+  protected:
+    template <typename THeadFn, typename TTailFn>
+    void ForEachEdge(RoadInfo const & roadInfo, THeadFn && headFn, TTailFn && tailFn)
```

Comment:
ForEachEdge is implemented as void ForEachEdge(FeatureID const & id, RoadInfo const & info, TFn && fn)